### PR TITLE
logging: add config dynamic reloading support

### PR DESCRIPTION
**Pull request by lxbsz**
Using this patch there is no need to restart the gluster-blockd
daemon after the /etc/sysconfig/gluster-blockd is changed.

Currently we only support the logging dynamic reloading, and the
others config parameters will be supported later.

System config for gluster-block, for now there are only 3 option types supported:
1, The "int type" option, for example:
   gb_int = 2

2, The "string type" option, for example:
     gb_str = "Tom"  --> Tom
   or
     gb_str = 'Tom'  --> Tom
   or
     gb_str = 'Tom is a "boy"' ---> Tom is a "boy"
   or
     gb_str = "'T' is short for Tom" --> 'T' is short for Tom

3, The "boolean type" option, for example:
   gb_bool

Fixes: #61

Signed-off-by: Xiubo Li <xiubli@redhat.com>

### CODE REVIEW

**Comment by lxbsz** [comment]
@pkalever @pranithk 

Last night I spent 3 hours for adding the config dynamic reloading support, and most of the code is from the tcmu-runner project. Please review.

Thanks,

**Comment by lxbsz** [comment]
@pkalever update this against to the upstream code.

File: utils/dyn-config.c [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+  return cfg;
+
+freeConfigPath:
+  free(cfg->configPath);
+freeConfig:
+  free(cfg);
+  return NULL;
+}
+
+static void glusterBlockCancelConfigThread(struct gbConfig *cfg)
```
Comment (pkalever) [inline utils/dyn-config.c]: general style here is:

```
returnType
functionName()
{
  logic;
  ...
}

```

**Comment by lxbsz** [comment]
@pkalever 

Yeah, all are good catches and will fix them.

Thanks,

File: utils/dyn-config.c [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+          struct gbConfOption *option; \
+          char buf[1024]; \
+          option = glusterBlockGetOption(#key); \
+          if (option) { \
+            if (cfg->key) \
+              GB_FREE(cfg->key); \
+            GB_STRDUP(cfg->key, option->optStr); \
+            if (option->optStr) \
+              GB_FREE(option->optStr); \
+            sprintf(buf, "%s", def); \
```
Comment (pkalever) [inline utils/dyn-config.c]: Please use a safe variant snprintf()

File: utils/dyn-config.c [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+static void glusterBlockConfFreeStrKeys(struct gbConfig *cfg)
+{
+  /* add your str type config options
+   *
+   * For example:
+   * GB_FREE_CFG_STR_KEY(cfg, 'STR KEY');
+   */
+}
+
+#define GB_MAX_CFG_FILE_SIZE (2 * 1024 * 1024)
```
[the remark this answers is not in the document]
Comment (pkalever) [reply utils/dyn-config.c]: This is still a question for me :-)

File: utils/dyn-config.c [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+    return NULL;
+  }
+
+  LOG("dyncfg", GB_LOG_INFO,
+      "Inotify is watching \"%s\", wd: %d, mask: IN_ALL_EVENTS\n",
+      cfg->configPath, wd);
+
+  while (1) {
+    struct inotify_event *event;
+    char *p;
```
Comment (pkalever) [inline utils/dyn-config.c]: can we move this to beginning of the function ?

File: utils/dyn-config.c [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+  char buf[BUF_LEN];
+
+  monitor = inotify_init();
+  if (monitor == -1) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to init inotify %d\n", monitor);
+    return NULL;
+  }
+
+  wd = inotify_add_watch(monitor, cfg->configPath, IN_ALL_EVENTS);
```
Comment (pkalever) [inline utils/dyn-config.c]: I think masking IN_MODIFY should be good enough, why do we need watch for IN_ALL_EVENTS ?

**Comment by pkalever** [comment]
@lxbsz
The other two patches looks good to me. Once you are done with the above changes/comments, I will test this.

Thanks!

File: utils/dyn-config.c (hunk 5 above)
Comment (lxbsz) [reply utils/dyn-config.c]: Yeah, there was one story about this when testing this in tcmu-runner's dynamic reloading, long time ago, but currently I couldn't recall it. Just could remember it won't work in some corner case.

File: utils/dyn-config.c (hunk 4 above)
Comment (lxbsz) [reply utils/dyn-config.c]: Yeah, will fix it.

#define GB_MAX_CFG_FILE_SIZE (2 * 1024 * 1024)

Sorry just forgot this one, the size is just calculated the possible size including the commented lines in the /etc/sysconfig/gluster-blockd files.  Or do you have any other ideas about this ? And I am okay with other reasonable size.

File: utils/dyn-config.c (hunk 2 above)
Comment (lxbsz) [reply utils/dyn-config.c]: sure.

File: utils/dyn-config.c (hunk 4 above)
Comment (pkalever) [reply utils/dyn-config.c]: Read one line by line and move some logic about checking for valid config line (skip blank, commented lines) from glusterBlockParseOptions() to glusterBlockReadConfig(), then sort/fetch the gb_options[] in glusterBlockReadConfig(), then call glusterBlockConfSetOptions() finally :-)

File: utils/dyn-config.c [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+}
+
+static int
+glusterBlockLoadConfig(gbConfig *cfg, bool reloading)
+{
+  int ret = -1;
+  int fd, len;
+  char *buf;
+
+  buf = CALLOC(GB_MAX_CFG_FILE_SIZE);
```
Comment (pkalever) [inline utils/dyn-config.c]: Use GB_ALLOC() instead of CALLOC()

GB_ALLOC() will internally uses calloc()

**Comment by pkalever** [comment]
Probably my earlier question got missed somewhere, let me ask again,

do we really need a separate gluster-block-general.log file ? can't we use gluster-blockd.log for this ?

Also you naming has a typo:
s/gluster-block-genernal.log/gluster-block-general.log/

File: utils/dyn-config.c (hunk 6 above)
Comment (lxbsz) [reply utils/dyn-config.c]: This won't work, the buf here is one 'char *' pointer, not struct type pointer.

Sep 20 21:35:59 rhel3 gluster-blockd[7532]: *** Error in `/usr/local/sbin/gluster-blockd': free(): invalid next size (fast): 0x0000000000db9070 ***

File: utils/dyn-config.c (hunk 4 above)
Comment (lxbsz) [reply utils/dyn-config.c]: Yeah, this make sense.

Do you mind to do a fix in the following PR ?

I am worried the stable of the dyn feature. Since this code is from the tcmu-runner and has been used and test for a very long time, to address this it will change a lot.

If today is not the deadline, I can change this and need to have long time test very possible corner cases.

Comment (pkalever) [reply utils/dyn-config.c]: I'm okay, fixing this later, but we are allocating 2 * 1024 * (1024 bytes) = 2 2MiB sole for this. Isn't that too big ? can we atleast shrink this ?

File: utils/dyn-config.c (hunk 6 above)
Comment (pkalever) [reply utils/dyn-config.c]: What ?

Did you use:
```
if (GB_ALLOC_N(buf, GB_MAX_CFG_FILE_SIZE) < 0) {
...
}
```

**Comment by pkalever** [comment]
> Before I thought the dyn logger is not only called by gluster-blockd daemon, but also cli and all the others. While the dyn logger is as one daemon service and I am okay use the gluster-blockd.log here.

@lxbsz can you paste few logs  of dyncfg ? depending on the amount of logs we can take a decision, may be.

File: utils/dyn-config.c (hunk 6 above)
Comment (lxbsz) [reply utils/dyn-config.c]: Okay, will use this. :-)

**Comment by lxbsz** [comment]
> @lxbsz can you paste few logs of dyncfg ? depending on the amount of logs we can take a decision, may be.

The logs are very simple and currently only the dyncfg will use this, I just removed the code.
The logs are like:

/var/log/gluster-block/gluster-blockd.log-general:275: [at lru.c+44 :<glusterBlockSetLruCount>]
/var/log/gluster-block/gluster-blockd-general.log:300:[2018-09-21 01:25:19.867594] INFO: glfsLruCount now is 31

Thanks,

File: utils/dyn-config.c (hunk 4 above)
Comment (lxbsz) [reply utils/dyn-config.c]: Yeah, let's use 32K for now, is that okay ?

Comment (pkalever) [reply utils/dyn-config.c]: Yeah, that should be more than sufficient IMO.

Comment (lxbsz) [reply utils/dyn-config.c]: Sure, done.

**Comment by pkalever** [comment]
> > @lxbsz can you paste few logs of dyncfg ? depending on the amount of logs we can take a decision, may be.
> 
> The logs are very simple and currently only the dyncfg will use this, I just removed the code.
> The logs are like:
> 
> /var/log/gluster-block/gluster-blockd.log-general:275: [at lru.c+44 :]
> /var/log/gluster-block/gluster-blockd-general.log:300:[2018-09-21 01:25:19.867594] INFO: glfsLruCount now is 31

Yeah, this looks simple, lets use gluster-blockd.log itself.

Thanks!

**Comment by lxbsz** [comment]
Yeah, all are updated. @pkalever

File: utils/lru.c [hunk 7]
```diff
@@ -22,6 +22,29 @@ typedef struct Entry {
...
+    MSG("glfsLruCount should be [0 < COUNT < %d]\n",
+        LRU_COUNT_MAX);
+    LOG(dom, GB_LOG_ERROR,
+        "glfsLruCount should be [0 < COUNT < %d]\n",
+        LRU_COUNT_MAX);
+    return -1;
+  }
+
+  LOCK(lru_lock);
+  glfsLruCount = lruCount;
```
Comment (pkalever) [inline utils/lru.c]: Is it a good idea to guard log-level too ?

File: daemon/gluster-blockd.c [hunk 8]
```diff
@@ -333,9 +334,15 @@ main (int argc, char **argv)
     exit(EXIT_FAILURE);
   }
 
+  gbCfg = glusterBlockSetupConfig(NULL);
+  if (!gbCfg) {
+    LOG("mgmt", GB_LOG_ERROR, "%s", "glusterBlockSetupConfig() failed");
+    return -1;
+  }
+
   if (glusterBlockDParseArgs(argc, argv)) {
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply daemon/gluster-blockd.c]: Yeah, this still the same as before:
1, when the options are commented in /etc/sysconfig/gluster-blockd, it will use the value in gluster-blockd.service's:

 10 Environment="GB_GLFS_LRU_COUNT=5"
 11 Environment="GB_LOG_LEVEL=INFO"

And in /etc/sysconfig/gluster-blockd:
 # 10 GB_GLFS_LRU_COUNT=115                                                                                                                                                    
 # 16 GB_LOG_LEVEL=TRACE

● gluster-blockd.service - Gluster block storage utility
   Loaded: loaded (/usr/local/lib/systemd/system/gluster-blockd.service; enabled; vendor preset: disabled)
   Active: active (running) since Fri 2018-09-21 01:11:58 EDT; 4s ago
 Main PID: 11064 (gluster-blockd)
   CGroup: /system.slice/gluster-blockd.service
           └─11064 /usr/local/sbin/gluster-blockd --glfs-lru-count 5 --log-level INFO


2, if we uncomment the options in /etc/sysconfig/gluster-blockd, it will use the values in in gluster-blockd.service instead:

 10 GB_GLFS_LRU_COUNT=115                                                                                                                                                    
 16 GB_LOG_LEVEL=TRACE

● gluster-blockd.service - Gluster block storage utility
   Loaded: loaded (/usr/local/lib/systemd/system/gluster-blockd.service; enabled; vendor preset: disabled)
   Active: active (running) since Fri 2018-09-21 01:15:28 EDT; 2s ago
 Main PID: 11162 (gluster-blockd)
   CGroup: /system.slice/gluster-blockd.service
           └─11162 /usr/local/sbin/gluster-blockd --glfs-lru-count 115 --log-level TRACE

[the remark this answers is not in the document]
Comment (pkalever) [reply daemon/gluster-blockd.c]: Okay. That should be good.

**Comment by pkalever** [comment]
LGTM - I will have a test and merge.

@lxbsz did you already test the latest version ?

Thanks!

**Comment by lxbsz** [comment]
@pkalever 
Yeah, already rebased against the newest code and test it.

**Comment by pkalever** [comment]
Looks good to me.

Tested-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>
Reviewed-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

Xiubo, please raise a new github issue for reading config file line by line instead of allocating a 32K buffer and address later.

Thanks @lxbsz

**Comment by lxbsz** [comment]
@pkalever Sure, thanks,

**Comment by lxbsz** [comment]
@pkalever 
Done, thanks.

**Comment by pkalever** [comment]
Thanks @lxbsz.
